### PR TITLE
[Yul] Deterministic YulStringRepository using string hashes.

### LINKED ISSUE #5378: [Yul] Explicit lifetime for YulStringRepository
Although #5377 seems to fix the immediate non-determinism issue we had, I think we should revisit and improve this at some point.

I think it would be cleaner to have the ``YulStringRepository`` be a member to some kind of context class (that doesn't exist at the time of writing) or a local variable somewhere or to give it an explicit lifetime in some other way. There could still be a static pointer (or ``std::weak_ptr``?) to this global object in ``YulStringRepository`` that can be initialized upon creation and uninitialized/set to zero upon destruction of the ``YulStringRepository``. ``YulStringRepository::instance()`` should assert that a ``YulStringRepository`` is currently available and return it, but throw if there is no global ``YulStringRepository`` instead of just creating one.

### PR DESCRIPTION
**Pull request by ekpyron**
Fixes #5378.

### CODE REVIEW

**Comment by ekpyron** [comment]
@chriseth What do you think about this approach conceptually? This has slightly worse performance characteristics, but should always be deterministic.
I tried giving the ``YulStringRepository`` an explicit lifetime instead on the ``yulStringRepoLifetime``-branch, but in comparison it's a mess and not really reliable...

File: libyul/YulString.h [hunk 1]
```diff
@@ -35,69 +35,84 @@ namespace yul
...
+			hash ^= *it;
+		}
 
+		return hash;
+	}
+	static constexpr std::uint64_t emptyHash()	{ return 14695981039346656037u; }
 private:
 	std::vector<std::shared_ptr<std::string>> m_strings;
-	std::map<std::string, size_t> m_ids;
+	std::unordered_multimap<std::uint64_t, size_t> m_hashToID;
```
Comment (ekpyron) [inline libyul/YulString.h:87]: Maybe a ``std::multimap`` is sufficient.

**Comment by ekpyron** [comment]
Tests are (probably) failing due to changed string order.

File: libyul/YulString.h [hunk 2]
```diff
@@ -35,69 +35,84 @@ namespace yul
...
-	{
-		return *m_strings.at(_id);
+			return { 0, emptyHash() };
+		std::uint64_t h = hash(_string);
+		auto range = m_hashToID.equal_range(h);
+		for (auto it = range.first; it != range.second; ++it)
+			if (*m_strings[it->second] == _string)
+				return Handle{it->second, h};
+		m_strings.emplace_back(std::make_shared<std::string>(_string));
+		size_t id = m_strings.size() - 1;
```
Comment (chriseth) [inline libyul/YulString.h:66]: Good idea, but doesn't the allocation of IDs still depend on the insertion order, at least for strings with the same hash?
Also, if we use hashes, why not just let `std::unordered_map<string, id>` handle the hash collisions for us?

Comment (chriseth) [reply libyul/YulString.h:66]: I still don't fully understand the pros and cons here, let's talk about it later!

**Comment by ekpyron** [comment]
@chriseth Should be ready for another review (I'm not exactly sure what to put in the comments actually...).

**Comment by ekpyron** [comment]
I created #5401 to track evaluating and/or improving the performance of this.

**Review by chriseth — APPROVED** [review]

**Pull request merged**